### ISSUE TITLE
busy error log spamming with unreachable kafka node

### ISSUE THREAD

**Issue by MartinNowak** [zendesk/ruby-kafka#288]
If this is a bug report, please fill out the following:
- Version of Ruby: 2.2.4
- Version of Kafka: 0.9.0.1
- Version of ruby-kafka: 0.3.15

Please verify that the problem you're seeing hasn't been fixed by the current `master` of ruby-kafka.
###### Steps to reproduce

``` ruby
kafka = Kafka.new(
  seed_brokers: ['kafka1:9092', 'kafka2:9092', 'kafka3:9092'],
  logger: logger)
consumer.subscribe('topic')
consumer.each_batch(min_bytes: 64 * 1024, max_wait_time: 5) do |kafka_batch|
  kafka_batch.messages.each do |msg|
    # process message
  end
end
```
###### Expected outcome

When one of the Kafka nodes went down, the consumer should have logged an error and tried to reconnect to the cluster/find the current coordinator.
###### Actual outcome

While restarting one of the Kafka replicas (might have been the leader for the partition), the consumer spammed the error log w/ about 2 messages / 4 ms (500 msg / s).

```
2016-10-20 02:17:40.753 cassandra-writer
Connection error while fetching messages: Connection refused - connect(2) for 10.1.24.5:9092
2016-10-20 02:17:40.749 cassandra-writer
Failed to connect to kafka3:9092: Connection refused - connect(2) for 10.1.24.5:9092
2016-10-20 02:17:40.749 cassandra-writer
Connection error while fetching messages: Connection refused - connect(2) for 10.1.24.5:9092
2016-10-20 02:17:40.745 cassandra-writer
Failed to connect to kafka3:9092: Connection refused - connect(2) for 10.1.24.5:9092
2016-10-20 02:17:40.745 cassandra-writer
Connection error while fetching messages: Connection refused - connect(2) for 10.1.24.5:9092
2016-10-20 02:17:40.741 cassandra-writer
Failed to connect to kafka3:9092: Connection refused - connect(2) for 10.1.24.5:9092
2016-10-20 02:17:40.741 cassandra-writer
Connection error while fetching messages: Connection refused - connect(2) for 10.1.24.5:9092
```

The error log end with this before the consumer apparently managed to reconnect to the restarted server or another one.

```
2016-10-20 02:17:46.167 cassandra-writer
Failed to find coordinator for group `cassandra_writer`; retrying...
2016-10-20 02:17:46.138 cassandra-writer
Error committing offsets: Kafka::NotCoordinatorForGroup
```

**Comment by dasch** [comment]
This sounds like something we've run into in the past. I'll try to take another stab at it when I'm back from leave in December.

**Comment by mensfeld** [comment]
Same issue here

**Comment by dasch** [comment]
I'm not quite sure what the best approach here would be – there are several options. The partitions owned by the broker could be paused for some duration, or there could be a simple sleep. I'd need to look more into this.

**Comment by mensfeld** [comment]
@dasch  hello after a long break :) I can provide you with a working example of that issue in 1 - 2 weeks if you want.

**Comment by dasch** [comment]
@mensfeld that would be great! Handling backoffs correctly is pretty tricky. Unless there's a functional impact I'm reluctant to touch it, but perhaps in this case it would be a good idea – hitting the broker with that many connection attempts is probably bad.

**Comment by mensfeld** [comment]
@dasch  @MartinNowak  were you guys able to reproduce this? I've tried with both docker setups that I've mentioned before and with 0.3.16  works great. If none can reproduce and no new people are submitting same issue, maybe this could be closed?

**Comment by rhinon** [comment]
I'm experiencing the same issue with 0.3.16 currently under different circumstances and versions.

If I have a very basic setup locally (1 kafka instance) and I stop that kafka instance, I get similar log spam. When I start the kafka instance back up, a `Kafka::Error` is thrown with "Failed to find group coordinator".

* Version of Ruby: 2.2.3
* Version of Kafka: 0.10.1.0
* Version of ruby-kafka: 0.3.16

**Comment by rhinon** [comment]
Another behavioral note: when it is in this state (Kafka unreachable loop), calling `stop` does not do anything. It will remain in an endless loop until it can reconnect.

**Comment by mensfeld** [comment]
@rhinon  could you create a reproduction procedure including the script you run? This would be helpful as I'm trying to reproduce it as well.

**Comment by rhinon** [comment]
@mensfeld This is an excerpt of my code which runs in the context of a daemon ruby process.
```ruby
@kafka = Kafka.new(
  logger: Rails.logger,
  seed_brokers: @seed_brokers
)

@kafka_consumer = @kafka.consumer(
  group_id: @consumer_group_id,
  offset_commit_interval: 10,
  offset_commit_threshold: 1
)

@kafka_consumer.subscribe(@topic, start_from_beginning: false)

@kafka_consumer.each_message do |message|
  Rails.logger.info "Received message key=#{message.key} on topic=#{message.topic}" +
    " partition=#{message.partition} offset=#{message.offset}"

  begin
    process_message(message.value)
  rescue => error
    Rails.logger.error "error=\"#{error.class}\" message=\"#{error.message}\""
    error.backtrace.each { |line| Rails.logger.error line }
  end

  Rails.logger.info "Finished processing message key=#{message.key}"
end
```

With a single Kafka process running locally, I stop that Kafka process (via ctrl+c) and I immediately get rapid looping logs:
```
I, [2017-03-23T18:22:27.960411 #38385]  INFO -- : Joining group `my.kafka.topic`
I, [2017-03-23T18:22:27.960504 #38385]  INFO -- : Fetching cluster metadata from kafka://localhost:9092
E, [2017-03-23T18:22:27.961334 #38385] ERROR -- : Failed to connect to localhost:9092: Connection refused - connect(2) for [::1]:9092
E, [2017-03-23T18:22:27.961439 #38385] ERROR -- : Failed to fetch metadata from kafka://localhost:9092: Connection refused - connect(2) for [::1]:9092
E, [2017-03-23T18:22:27.961524 #38385] ERROR -- : Connection error while trying to join group `primary.jhl7.mbmc`; retrying...
```

While in this state, calling `@kafka_consumer.stop` does not stop the looping.

If I start the Kafka process again it will raise this error and break the run loop (note that my exception log here is mixed in with the logs from ruby-kafka):
```
I, [2017-03-23T18:22:29.920725 #38385]  INFO -- : Joining group `my.kafka.topic`
I, [2017-03-23T18:22:29.920792 #38385]  INFO -- : Fetching cluster metadata from kafka://localhost:9092
I, [2017-03-23T18:22:30.516299 #38385]  INFO -- : Discovered cluster metadata; nodes:
I, [2017-03-23T18:22:30.516553 #38385]  INFO -- : Leaving group `my.kafka.topic`
E, [2017-03-23T18:22:30.516693 #38385] ERROR -- : Caught error while kafka looping!
E, [2017-03-23T18:22:30.516809 #38385] ERROR -- : error="Kafka::Error" message="Failed to find group coordinator"
E, [2017-03-23T18:22:30.516963 #38385] ERROR -- : /Users/rhinon/.rvm/gems/ruby-2.2.3/gems/ruby-kafka-0.3.16/lib/kafka/cluster.rb:110:in `get_group_coordinator'
E, [2017-03-23T18:22:30.517064 #38385] ERROR -- : /Users/rhinon/.rvm/gems/ruby-2.2.3/gems/ruby-kafka-0.3.16/lib/kafka/consumer_group.rb:164:in `coordinator'
E, [2017-03-23T18:22:30.517157 #38385] ERROR -- : /Users/rhinon/.rvm/gems/ruby-2.2.3/gems/ruby-kafka-0.3.16/lib/kafka/consumer_group.rb:108:in `join_group'
E, [2017-03-23T18:22:30.517249 #38385] ERROR -- : /Users/rhinon/.rvm/gems/ruby-2.2.3/gems/ruby-kafka-0.3.16/lib/kafka/consumer_group.rb:39:in `join'
E, [2017-03-23T18:22:30.517337 #38385] ERROR -- : /Users/rhinon/.rvm/gems/ruby-2.2.3/gems/ruby-kafka-0.3.16/lib/kafka/consumer.rb:303:in `join_group'
E, [2017-03-23T18:22:30.517425 #38385] ERROR -- : /Users/rhinon/.rvm/gems/ruby-2.2.3/gems/ruby-kafka-0.3.16/lib/kafka/consumer.rb:284:in `rescue in consumer_loop'
E, [2017-03-23T18:22:30.517512 #38385] ERROR -- : /Users/rhinon/.rvm/gems/ruby-2.2.3/gems/ruby-kafka-0.3.16/lib/kafka/consumer.rb:281:in `consumer_loop'
E, [2017-03-23T18:22:30.517598 #38385] ERROR -- : /Users/rhinon/.rvm/gems/ruby-2.2.3/gems/ruby-kafka-0.3.16/lib/kafka/consumer.rb:176:in `each_message'
<cut out the rest of the stack trace that doesn't seem important>
I, [2017-03-23T18:22:30.520361 #38385]  INFO -- : Finished run-loop, exiting.
```
Also note that this happens whether or not I call `@kafka_consumer.stop` while kafka is down.


I have some suspicions but I will probably not have time to debug and deep dive into the `ruby-kafka` code until this weekend.

**Comment by mensfeld** [comment]
I'm able to reproduce now:

1. Connect to Kafka node
2. Make it connect etc
3. Kill -9 everything (I mean kafka process) :D

Won't stop connection attempts

**Comment by mensfeld** [comment]
FYI this is how it drains disk space when it keeps spamming
![df-week](https://user-images.githubusercontent.com/392754/28174431-cc1e993e-67f1-11e7-92d0-fadc88d9e4bf.png)

**Comment by dasch** [comment]
@mensfeld sounds like a sleep in https://github.com/zendesk/ruby-kafka/blob/master/lib/kafka/consumer_group.rb#L48 would help?

**Comment by mensfeld** [comment]
@dasch exactly :+1:

**Comment by dasch** [comment]
Should be fixed by #288.
